File: msg-sentiment/src/test/java/dev/number6/sentiment/dagger/FakeComprehensionResultsModule.java
package dev.number6.sentiment.dagger;

import dev.number6.comprehend.port.ComprehensionPort;
import dev.number6.comprehend.results.PresentableSentimentResults;
import dev.number6.db.port.DatabasePort;
import dev.number6.message.ChannelMessages;
import dev.number6.message.ChannelMessagesToComprehensionResultsFunction;
import dev.number6.message.ComprehensionResultsConsumer;
import dagger.Module;
import dagger.Provides;

import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.fail;

@Module
public class FakeComprehensionResultsModule {

    @Provides
    @Singleton
    public ChannelMessagesToComprehensionResultsFunction<PresentableSentimentResults> providesMessageToSentimentResults(ComprehensionPort comprehensionPort) {
        return new ConfigurableResultsFunction();
    }

    @Provides
    @Singleton
    public ComprehensionResultsConsumer<PresentableSentimentResults> providesSentimentResultsConsumer(DatabasePort databasePort) {
        return new RecordingComprehensionResultsConsumer();
    }

    public static class ConfigurableResultsFunction implements ChannelMessagesToComprehensionResultsFunction<PresentableSentimentResults> {

        private PresentableSentimentResults presentableSentimentResults;

        @Override
        public PresentableSentimentResults apply(ChannelMessages channelMessages) {
            return presentableSentimentResults == null ?
                    fail("presentable entity result needs to be configured") :
                    presentableSentimentResults;
        }

        public void setPresentableSentimentResults(PresentableSentimentResults presentableSentimentResults) {
            this.presentableSentimentResults = presentableSentimentResults;
        }
    }

    public static class RecordingComprehensionResultsConsumer implements ComprehensionResultsConsumer<PresentableSentimentResults> {
        private List<PresentableSentimentResults> resultsConsumed = new ArrayList<>();

        public List<PresentableSentimentResults> getResultsConsumed() {
            return resultsConsumed;
        }

        @Override
        public void accept(PresentableSentimentResults presentableSentimentResults) {
            resultsConsumed.add(presentableSentimentResults);
        }
    }
}

File: msg-sentiment/src/test/kotlin/dev/number6/sentiment/dagger/FakeComprehensionResultsModule.kt
package dev.number6.sentiment.dagger

import dagger.Module
import dagger.Provides
import dev.number6.comprehend.port.ComprehensionPort
import dev.number6.comprehend.results.PresentableSentimentResults
import dev.number6.db.port.DatabasePort
import dev.number6.message.ChannelMessages
import dev.number6.message.ChannelMessagesToComprehensionResultsFunction
import dev.number6.message.ComprehensionResultsConsumer
import org.junit.jupiter.api.Assertions
import java.util.*
import javax.inject.Singleton

@Module
class FakeComprehensionResultsModule {
    @Provides
    @Singleton
    fun providesMessageToSentimentResults(): ChannelMessagesToComprehensionResultsFunction<PresentableSentimentResults> {
        return ConfigurableResultsFunction()
    }

    @Provides
    @Singleton
    fun providesSentimentResultsConsumer(): ComprehensionResultsConsumer<PresentableSentimentResults> {
        return RecordingComprehensionResultsConsumer()
    }

    class ConfigurableResultsFunction : ChannelMessagesToComprehensionResultsFunction<PresentableSentimentResults> {
        private var presentableSentimentResults: PresentableSentimentResults? = null
        override fun apply(channelMessages: ChannelMessages): PresentableSentimentResults {
            return presentableSentimentResults ?: Assertions.fail("presentable entity result needs to be configured")
        }

        fun setPresentableSentimentResults(presentableSentimentResults: PresentableSentimentResults) {
            this.presentableSentimentResults = presentableSentimentResults
        }
    }

    class RecordingComprehensionResultsConsumer : ComprehensionResultsConsumer<PresentableSentimentResults> {
        private val resultsConsumed: MutableList<PresentableSentimentResults> = ArrayList()
        fun getResultsConsumed(): MutableList<PresentableSentimentResults> {
            return resultsConsumed
        }

        override fun accept(presentableSentimentResults: PresentableSentimentResults) {
            resultsConsumed.add(presentableSentimentResults)
        }
    }
}